n3fit/src/n3fit/model_trainer.py

```diff
@@ -327,15 +326,39 @@ def _model_generation(self, pdf_models, partition, partition_idx):
         """
         log.info("Generating the Model")
 
-        # Construct the input array that will be given to the pdf
-        input_arr = np.concatenate(self.input_list, axis=1).T
+        # In the case of pineappl models all fktables ask for the same grid in x
+        # and so the input can be simplified to be a single grid for all dataset
+        # instead of a concatenation that gets splitted afterwards
+        # However, this is not a _strict_ requirement for pineappl so the solution below
+        # aims to be completely general
+        # Detailed:
+        #    let's assume an input [x1, x1, x1, x2, x2, x3]
+        #    where each xi is a different grid, this will be broken into two lists:
+        #    [x1, x2, x3] (unique grids) and [0,0,0,1,1,2] (index of the grid per dataset)
+        #    The pdf will then be evaluated to concatenate([x1,x2,x3]) and then split (x1, x2, x3)
+        #    Then each of the experiment, looking at the indexes, will receive one of the 3 PDFs
+
+        # Clean the input list
+        inputs_hash = []
+        inputs_unique = []
+        inputs_idx = []
+        for input_grid in self.input_list:
+            ghash = hash(tuple(input_grid.flatten()))
+            if ghash not in inputs_hash:
+                inputs_hash.append(ghash)
+                inputs_unique.append(input_grid)
+            inputs_idx.append(inputs_hash.index(ghash))
```

Comment:
```suggestion
        inputs_unique = []
        inputs_idx = []
        for igrid in self.input_list:
            for idx, arr in enumerate(inputs_unique):
                if np.array_equal(igrid, arr):
                    inputs_idx.append(idx)
                    break
            else:
                inputs_idx.append(len(inputs_unique))
                inputs_unique.append(igrid)
```
This should work. And allows for some fine-tunning on what the equality should be, which I like. 

Although it becomes even more complicated.

Reply:
I think this is pretty clear, and the preferable way if you want to use allclose. 

Hashes would also be fine with me (if you don't care about allclose), of course they're technically not unique as Zahari pionts out, but in practice we'd have to work pretty hard to find two objects returning the same hash. 

 I don't really like the idea of checking for default names inside n3fit. That's pineappl business.

Reply:
There is no reason why you would use hashes but not dicts or sets which use hashes and additionally an equality check (quick identity check in the happy case). I think it is best to use none of that.